lib/unparser/emitter/literal/primitive.rb

```diff
@@ -7,10 +7,41 @@ class Primitive < self
 
         children :value
 
+        # Emitter for string literals
+        class String < self
+          handle :str
+
+        private
+
+          # Dispatch value
+          #
+          # @return [undefined]
+          #
+          # @api private
+          #
+          def dispatch
+            if value.include? "'"
+              # Write %()-style strings for strings with both double and single
+              # quotes
+              if value.include? '"'
+                write("%(#{value})")
```

Comment:
As there are alternative escape characters, I think this whole branch might best be factored out into a private method, trying them in preference order.